### PR TITLE
Zeng-Hu Hoverbeds

### PR DESCRIPTION
**Pull request by 1Sparky1**
Adds Zeng-Hu Hoverbeds, which are now less of a novelty and more practically helpful due to the stairs changes. They can be sourced from R&D, ~~and the First Responders start with two to share between themselves.~~

### CODE REVIEW

**Comment by 1Sparky1** [comment]
!review

File: code/modules/research/designs/protolathe/medical_designs.dm [hunk 1]
```diff
@@ -96,6 +96,13 @@
 	materials = list(DEFAULT_WALL_MATERIAL = 3000, MATERIAL_PHORON = 3000, MATERIAL_DIAMOND = 500)
 	build_path = /obj/item/reagent_containers/personal_inhaler_cartridge/bluespace
 
+/datum/design/item/medical/zeng_roller
+	name = "Zeng-Hu Hoverbed"
+	desc = "A medical hoverbed, designed by Zeng-Hu Pharmaceuticals in conjunction with the Nralakk Federation."
```
Comment (Greenjoe12345) [inline code/modules/research/designs/protolathe/medical_designs.dm:101]: Maybe this description should be added to the bed item itself too?

**Comment by FluffyGhoster** [comment]
I think the FRs starting with them basically circumvents the whole added mechanic

If you want to make them buildable eg. by science that's fine, but them being present at roundstart is a no from me

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 2]
```diff
@@ -575,14 +575,23 @@
 	makes_rolling_sound = FALSE
 	held_item = /obj/item/roller/hover
 	has_iv_light = FALSE
+	var/hover_color = LIGHT_COLOR_CYAN
```
Comment (FluffyGhoster) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:578]: As above, and also this variable needs a DMDoc

**Review by FluffyGhoster — CHANGES_REQUESTED** [review]

**Comment by 1Sparky1** [comment]
I don't think 2 beds circumvents the whole mechanic - Most FRs carry at least 2 to my knowledge, and these would be split between them. And anyone doing medical retrieval to fill in for a missing FR would have to deal with the stairs still.

I also don't think making them exclusively research is beneficial to anyone - Just printing things isn't in itself a particularly satisfying experience, and it would mean a science- and/or mining-less medical would suffer heavily (i.e. low pop would almost never get to use them)

**Comment by FluffyGhoster** [comment]
> I don't think 2 beds circumvents the whole mechanic - Most FRs carry at least 2 to my knowledge, and these would be split between them. And anyone doing medical retrieval to fill in for a missing FR would have to deal with the stairs still.

You can only transport a single patient at a time, so yes in essence it bypasses the mechanic, when you reach the GTR you unload the patient on another bed and go back out, thus bypassing the need to account for the stairs/elevators

> I also don't think making them exclusively research is beneficial to anyone - Just printing things isn't in itself a particularly satisfying experience, and it would mean a science- and/or mining-less medical would suffer heavily (i.e. low pop would almost never get to use them)

We have various things exclusively research [sic], the IMS for surgeons for example; R&D is supposed to make and give improvements and QoL things, if people have fancy things by default, that reduces the utility of having R&D, thus lowering the amount of people that want to play it, and the loop continues

**Comment by 1Sparky1** [comment]
While I'm still of two minds, I'd rather see the item in game soon and have the debate over starting beds or not at another time, so I've removed the initial ones.

**Review by FluffyGhoster — APPROVED** [review]

**Review by DreamySkrell — APPROVED** [review]

**Review by Arrow768 — CHANGES_REQUESTED** [review]
Given that they (obviously) use power, they should have inbuilt powercells and all the mechanics that come with it:
- Adding/Removing the powercell
- EMP-able -> Discharges the powercell
- Hover function shutting down (ideally abruptly) if power runs out
- A warning before the power runs out

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 3]
```diff
@@ -576,13 +576,126 @@
...
+
 /obj/structure/bed/roller/hover/Initialize()
 	.=..()
-	set_light(2,1,LIGHT_COLOR_CYAN)
+	set_light(2,1,hover_color)
+	if(!cell || !cell.charge)
+		hover = FALSE
+	START_PROCESSING(SSprocessing, src)
+
+/obj/structure/bed/roller/hover/Destroy()
```
Comment (FluffyGhoster) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:611]: Need to STOP_PROCESSING(SSprocessing, src) here, probably

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 4]
```diff
@@ -576,13 +576,126 @@
...
+	QDEL_NULL(cell)
+	return ..()
+
+// Same as parent, but doesn't kill processing (We need that for power drain)
+/obj/structure/bed/roller/hover/detach_iv(mob/living/carbon/human/target, mob/user)
+	user.visible_message(SPAN_NOTICE("<b>[user]</b> takes [target] off the IV on \the [src]."), SPAN_NOTICE("You take the IV off \the [target]."))
+	iv_attached = FALSE
+	update_icon()
+
+/obj/structure/bed/roller/hover/process()
```
Comment (FluffyGhoster) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:621]: You need to respect the `ticks_per_second` parameter that it gives you, and use it for anything that works on time

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 5]
```diff
@@ -576,13 +576,126 @@
...
+	. = ..()
+	if(!hover && cell && cell.charge)
+		hover = TRUE
+		visible_message(SPAN_WARNING("\The [src] rises from the floor as its hover thrusters kick in."))
+		update_icon()
+	if(hover && cell)
+		cell.use(power_draw)
```
Comment (FluffyGhoster) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:628]: Like here, you need to multiply the power draw by the ticks_per_second

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 6]
```diff
@@ -576,13 +576,126 @@
...
+
+	if(!cell || !cell.charge)
+		hover = FALSE
+		visible_message(SPAN_WARNING("\The [src] falls the the floor as its hover thrusters cut out."))
```
Comment (FluffyGhoster) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:632]: Wouldn't this keep spamming out that it's out of charge? You are not stopping the processing below?

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 7]
```diff
@@ -576,13 +576,126 @@
...
+		update_icon()
```
Comment (FluffyGhoster) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:633]: Likewise, wouldn't this keep refreshing the icon?

**Review by FluffyGhoster — APPROVED** [review]
Looks good enough

**Pull request closed without merging**